kms/src/main/java/io/kroxylicious/kms/service/Kms.java

```diff
@@ -0,0 +1,104 @@
+/*
+ * Copyright Kroxylicious Authors.
+ *
+ * Licensed under the Apache Software License version 2.0, available at http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+package io.kroxylicious.kms.service;
+
+import java.util.concurrent.CompletionStage;
+
+import javax.crypto.SecretKey;
+
+import edu.umd.cs.findbugs.annotations.NonNull;
+
+/**
+ * Abstracts the KMS operations needed for Envelope Encryption
+ * @param <K> The type of Key Encryption Key id.
+ * @param <E> The type of encrypted Data Encryption Key.
+ */
+public interface Kms<K, E> {
```

Comment:
I'm not sure I fully grok the scope/intended usage of the KMS interface. 

My naive impression was the KMS was to abstract a key management service such as Hashicorp Vault or IBM key protect. Thus `generateDek` and `decryptEdek` make sense to me but I'm confused by the KMS exposing Serdes for encoding keyId's etc shouldn't that be internal to how the implementation communicates with the underlying KMS system?

Reply:
> I'm confused by the KMS exposing Serdes for encoding keyId's etc

We need to serialize kekIds and edeks into records to do envelope encryption. 

It _may_ be the case that all the KMSe we will (eventually need to) care about uniformly represent their key ids as `String` we can't know that for sure. It's also not certain whether there's really a natural representation of encrypted deks. By representing these are type parameters of `Kms` we have an abstraction that doesn't care. You can represent your edek as a (base64 encoded) `String` if you want, or a `byte[]` or a  `ByteBuffer`, or a `MyKmsEdekType`.  

Bonus: At the point of use, in the encryption filter, we have to parametrize things with `<K>` and `<E>`, which is a bit of a pain, but it also means the compiler sees them as distinct types which is more type safe than if both kekId and edek were `String`.

Reply:
> We need to serialize kekIds and edeks into records to do envelope encryption.

Argghhh I had this thought but apparently failed to include it. 
 

That's the step I don't follow, why is the serialisation to the record governed by the choice of key management service?

Reply:
In this model the KMS chooses the Java representation of it's keys and deks, and we have to serialize them to and fro in the filter. 
The alternative model that I think you're thinking of would have the KMSes do all the serialization internally to and from some common representation (say `KekId` and `Edek` classes) declared in the KMS module. 
Both models can be made to work. 

The former is slightly more flexible. For example, if you were using an SDK to implement the `Kms` abstraction and that returned a `FooKeyId` type then you have:
* `bytes from KMS service -> FooKeyId -> bytes in record` under the former model
* `bytes from KMS service -> FooKeyId -> KekId -> bytes in record` under the latter model

----

I think there's a point about key identity that needs more thought. It's orthogonal serialization, but it relates to the semantics we require for key ids... We serialize the key id along with the edek so that we can later ask the KMS to decrypt the DEK. This requires one of the following properties to hold:

1. that the KEK id we serialize cannot be changed to refer to another key later on. In other words it needs to identify a key "in space and time".
2. that the key id is _not actually needed_ for decrypting the DEK. [This is the case with AWS](https://docs.aws.amazon.com/kms/latest/APIReference/API_Decrypt.html), for example, where the binary blob of the edek already encodes the key id. This means the user can use a key alias to specify the KEK and even if they later change that alias the existing EDEKs will still be decryptable.   

Currently nothing in our KMS module makes this requirement clear, everything it just in terms of `K` and `E`.

Reply:
I think I'm looking to move the Key serde **out** of the KMS. 

I'm not questioning the need for a KMS to be able to serealize a keyId for lookup but that seems like an independent concern to what we choose to persist for book keeping. 

Now I could be completely wrong but as I understand the proposal the choice of how we track the key used in our book keeping is determined by the users choice of KMS implementation. That seems like the wrong reason for it to vary to me. Surely the book keeping format is the same if the user deploys the vault KMS or the AWS KMS? 

------

Inspired by your identity point. Do we have any need to support (in the long run) users being able to move keys between KMS implementations? Supporting migrations feels like the sort of thing users might well want given potential Kafka retention times however I'm not entirely sure its a cryptographically sane thing to do or if its really just too hairy to contemplate in practice.

I guess what I'm driving at does the KEK have an identity outside of a given KMS implementation? 

n.b. I really want a way to distinguish between our KMS abstraction and the underlying KMS system.

Reply:
> Surely the book keeping format is the same if the user deploys the vault KMS or the AWS KMS?

Wouldn't this require a KMS vendor neutral way to address a key, at a particular revision that is stable over time? Unfortunately there's no standard we can lean on.

Some KMSs support mutable aliases, but you wouldn't want to serialise based on those.  If the alias is pointed at a different key... bang!  There goes your decryptability guarantee.  

https://github.com/kroxylicious/design/discussions/42#user-content-key-naming


> Inspired by your identity point. Do we have any need to support (in the long run) users being able to move keys between KMS implementations? 

I don't think so.   Indeed, KMSs in general don't let you extract the key material.  IBM KeyProtect is an outlier. https://github.com/kroxylicious/design/discussions/42#user-content-export-key

Reply:
Hmm I'm starting to see the confusion. 

> Wouldn't this require a KMS vendor neutral way to address a key, at a particular revision that is stable over time? Unfortunately there's no standard we can lean on.

Right so in my head the key ref is effectively a relative path within the provided KMS or effectively a primary key. We could also store a "fully qualified" path including the address of the KMS. But either way to me that's portable if implementation specific.

Once we have the KMS specific keyref we then need to choose how to encode it, and it's that conversion I'm struggling to see why it needs to come from the KMS and not from the book keeper

Reply:
We said we weren't going to consider multiple KMS as in scope for the initial work. That definitely includes migration between KMS.

Reply:
> that the key id is not actually needed for decrypting the DEK. This is the case with AWS, for example, where the binary blob of the edek already encodes the key id. This means the user can use a key alias to specify the KEK and even if they later change that alias the existing EDEKs will still be decryptable.

I'm beginning to think this is the model we should enforce in the abstraction. I.e. we allow the kekId to be optional when unwrapping the edek. The forces the implementor to include the kekId in the edek. Which in turn means we don't need to specify stability semantics for the kekId.